File: sdk-java/src/main/java/com/battlesim/naval/tactical/Evader.java
package com.battlesim.naval.tactical;

import com.battlesim.naval.Command;
import com.battlesim.naval.protocol.TickEvent;
import com.battlesim.naval.protocol.WorldView;

import java.util.Optional;

/**
 * Hit-triggered evasive maneuver state machine.
 *
 * <p>Returns an override {@link Command} while evading; empty otherwise. The
 * orchestrator (or your own {@code onTick}) treats a non-empty return as a
 * preempting command — evasion outranks the player's tactical intent.
 */
public class Evader {

    public enum State { IDLE, EVADING, COOLDOWN }

    private final int evasionTicks;
    private final int cooldownTicks;
    private final float throttle;
    private State state = State.IDLE;
    private long stateUntil = 0;
    private float rudderSign;

    public Evader() {
        this(15, 10, 1.0f, 1.0f);
    }

    public Evader(int evasionTicks, int cooldownTicks) {
        this(evasionTicks, cooldownTicks, 1.0f, 1.0f);
    }

    public Evader(int evasionTicks, int cooldownTicks, float throttle, float initialRudderSign) {
        this.evasionTicks = evasionTicks;
        this.cooldownTicks = cooldownTicks;
        this.throttle = throttle;
        this.rudderSign = initialRudderSign >= 0 ? 1.0f : -1.0f;
    }

    public State state() { return state; }

    /**
     * Advance the state machine. Returns an override {@link Command} while
     * evading, else empty.
     */
    public Optional<Command> update(WorldView view) {
        long tick = view.tick();

        // Timed transitions.
        if (state != State.IDLE && tick >= stateUntil) {
            if (state == State.EVADING) {
                state = State.COOLDOWN;
                stateUntil = tick + cooldownTicks;
            } else {
                state = State.IDLE;
            }
        }

        // React to fresh hits.
        boolean wasHit = false;
        for (TickEvent e : view.events()) {
            if (e instanceof TickEvent.Hit) {
                wasHit = true;
                break;
            }
        }
        if (wasHit && (state == State.IDLE || state == State.COOLDOWN)) {
            if (state == State.COOLDOWN) {
                rudderSign = -rudderSign;
            }
            state = State.EVADING;
            stateUntil = tick + evasionTicks;
        }

        if (state == State.EVADING) {
            return Optional.of(new Command().throttle(throttle).rudder(rudderSign));
        }
        return Optional.empty();
    }
}

File: sdk-python/naval_sdk/tactical/evader.py
"""Hit-triggered evasive maneuver state machine.

Returns an override :class:`Command` while evading; ``None`` otherwise. The
orchestrator (:class:`TacticalBot`) treats a non-``None`` return as a
preempting command — evasion outranks the player's tactical intent.

See ``docs/design-decisions/sdk-tactical-toolkit.md`` §4.6.
"""

from __future__ import annotations

from enum import Enum
from typing import Optional

from ..protocol import Command, HitEvent, WorldView


class EvaderState(Enum):
    IDLE = "idle"
    EVADING = "evading"
    COOLDOWN = "cooldown"


class Evader:
    def __init__(
        self,
        *,
        evasion_ticks: int = 15,
        cooldown_ticks: int = 10,
        throttle: float = 1.0,
        initial_rudder_sign: float = 1.0,
    ) -> None:
        self._evasion_ticks = int(evasion_ticks)
        self._cooldown_ticks = int(cooldown_ticks)
        self._throttle = float(throttle)
        self._state: EvaderState = EvaderState.IDLE
        self._state_until: int = 0
        self._rudder_sign: float = 1.0 if initial_rudder_sign >= 0 else -1.0

    @property
    def state(self) -> EvaderState:
        return self._state

    def update(self, view: WorldView) -> Optional[Command]:
        """Advance the state machine. Returns an override ``Command`` while
        evading, else ``None``.
        """
        # Tick state transitions.
        if self._state != EvaderState.IDLE and view.tick >= self._state_until:
            if self._state == EvaderState.EVADING:
                self._state = EvaderState.COOLDOWN
                self._state_until = view.tick + self._cooldown_ticks
            else:
                self._state = EvaderState.IDLE

        # React to fresh hits.
        was_hit = any(isinstance(e, HitEvent) for e in view.events)
        if was_hit and self._state in (EvaderState.IDLE, EvaderState.COOLDOWN):
            if self._state == EvaderState.COOLDOWN:
                # Took another hit before we got clear — flip rudder to confuse aim.
                self._rudder_sign = -self._rudder_sign
            self._state = EvaderState.EVADING
            self._state_until = view.tick + self._evasion_ticks

        if self._state == EvaderState.EVADING:
            return Command(
                throttle=self._throttle,
                rudder=self._rudder_sign,
            )
        return None
